atlasexec/atlas_test.go

```diff
@@ -51,7 +59,180 @@ func Test_MigrateApply(t *testing.T) {
+		require.Nil(t, got)
+	})
+	t.Run("lint error parsing", func(t *testing.T) {
+		c, err := atlasexec.NewClient(".", "atlas")
+		require.NoError(t, err)
+		got, err := c.MigrateLint(context.Background(), &atlasexec.MigrateLintParams{
+			DevURL: "sqlite://file?mode=memory",
+			DirURL: "file://testdata/migrations",
+			Latest: 1,
+		})
+		require.NoError(t, err)
+		require.GreaterOrEqual(t, 4, len(got.Steps))
+		require.Equal(t, "sqlite3", got.Env.Driver)
+		require.Equal(t, "testdata/migrations", got.Env.Dir)
+		require.Equal(t, "sqlite://file?mode=memory", got.Env.URL.String())
+		require.Equal(t, 1, len(got.Files))
+		expectedReport := &x.FileReport{
+			Name: "20230926085734_destructive-change.sql",
+			Text: "DROP TABLE t2;\n",
+			Reports: []sqlcheck.Report{{
+				Text: "destructive changes detected",
+				Diagnostics: []sqlcheck.Diagnostic{{
+					Pos:  0,
+					Text: `Dropping table "t2"`,
+					Code: "DS102",
+				}},
+			}},
+			Error: "destructive changes detected",
+		}
+		require.EqualValues(t, expectedReport, got.Files[0])
+	})
+	t.Run("lint with manually parsing output", func(t *testing.T) {
+		c, err := atlasexec.NewClient(".", "atlas")
+		require.NoError(t, err)
+		var buf bytes.Buffer
+		err = c.MigrateLintError(context.Background(), &atlasexec.MigrateLintParams{
+			DevURL: "sqlite://file?mode=memory",
+			DirURL: "file://testdata/migrations",
+			Latest: 1,
+			Writer: &buf,
+		})
+		require.NoError(t, err)
+		var raw json.RawMessage
+		require.NoError(t, json.NewDecoder(&buf).Decode(&raw))
+	})
+}
+
+type graphQLQuery struct {
+	Query     string          `json:"query"`
+	Variables json.RawMessage `json:"variables"`
+}
+
+type Dir struct {
+	Name    string `json:"name"`
+	Content string `json:"content"`
+}
+type dirsQueryResponse struct {
+	Data struct {
+		Dirs []Dir `json:"dirs"`
+	} `json:"data"`
+}
+
+func TestMigrateLintWithLogin(t *testing.T) {
+	token := "123456789"
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		require.Equal(t, "Bearer "+token, r.Header.Get("Authorization"))
+		query := graphQLQuery{}
+		err := json.NewDecoder(r.Body).Decode(&query)
+		require.NoError(t, err)
+		switch {
+		case strings.Contains(query.Query, "mutation reportMigrationLint"):
+			fmt.Fprintf(w, `{ "data": { "reportMigrationLint": { "url": "https://migration-lint-report-url" } } }`)
+		}
+	}))
+	t.Cleanup(srv.Close)
+	atlasConfigURL := generateHCL(t, srv.URL, token)
+	t.Run("Web and Writer params produces an error", func(t *testing.T) {
+		c, err := atlasexec.NewClient(".", "atlas")
+		require.NoError(t, err)
+		params := &atlasexec.MigrateLintParams{
+			ConfigURL: atlasConfigURL,
+			DevURL:    "sqlite://file?mode=memory",
+			DirURL:    "file://testdata/migrations",
+			Latest:    1,
+		}
+		paramsWeb := *params
+		paramsWeb.Web = true
+		got, err := c.MigrateLint(context.Background(), &paramsWeb)
```

Comment:
why is that? just set params.Web.

Reply:
I don't want to treat the input as mutable, user passed a parameter by ref that is not (in my opinion) expected to change

Maybe pass by Value?

Reply:
This (java mindset) makes the test unpleasant to read - we also don’t maintain this invariant in ~anywhere in the codebase. 

Alternatively, you can define a non-pointer, copy it, set its field and pass the reference. But again, this is an unnecessary complexity

Reply:
Sorry, it's been a long day and I was commenting from my phone, I thought I was commenting on code that is inside the SDK (hence the comment about pass by value)

The reason for not just setting params.Web is because I would have to reset it afterwards and this way seemed cleaner to me